File: library/xml-model/src/main/java/org/cru/godtools/xml/model/Styles.java
package org.cru.godtools.xml.model;

import org.cru.godtools.xml.model.Text.Align;

import androidx.annotation.ColorInt;
import androidx.annotation.DimenRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

// XXX: I really want to use default methods for this interface, but default methods are not backwards compatible
public interface Styles {
    @ColorInt
    int getPrimaryColor();

    @ColorInt
    int getPrimaryTextColor();

    @ColorInt
    int getTextColor();

    @ColorInt
    int getButtonColor();

    @DimenRes
    int getTextSize();

    @NonNull
    Align getTextAlign();

    @ColorInt
    static int getPrimaryColor(@Nullable final Styles styles) {
        return styles != null ? styles.getPrimaryColor() : Manifest.getDefaultPrimaryColor();
    }

    @ColorInt
    static int getPrimaryTextColor(@Nullable final Styles styles) {
        return styles != null ? styles.getPrimaryTextColor() : Manifest.getDefaultPrimaryTextColor();
    }

    @ColorInt
    static int getTextColor(@Nullable final Styles styles) {
        return styles != null ? styles.getTextColor() : Manifest.getDefaultTextColor();
    }

    @ColorInt
    static int getButtonColor(@Nullable final Styles styles) {
        return styles != null ? styles.getButtonColor() : getPrimaryColor(null);
    }

    @DimenRes
    static int getTextSize(@Nullable final Styles styles) {
        return styles != null ? styles.getTextSize() : Manifest.getDefaultTextSize();
    }

    @NonNull
    static Align getTextAlign(@Nullable final Styles styles) {
        return styles != null ? styles.getTextAlign() : Manifest.getDefaultTextAlign();
    }
}

File: library/xml-model/src/main/java/org/cru/godtools/xml/model/Styles.kt
package org.cru.godtools.xml.model

import androidx.annotation.ColorInt
import org.cru.godtools.tool.model.buttonStyle
import org.cru.godtools.tool.model.primaryColor
import org.cru.godtools.tool.model.primaryTextColor
import org.cru.godtools.tool.model.textAlign
import org.cru.godtools.tool.model.textColor

internal const val DEFAULT_TEXT_SCALE = 1.0

@Deprecated("Use kotlin mpp library instead")
typealias Styles = org.cru.godtools.tool.model.Styles

@Deprecated("Use kotlin mpp library instead", ReplaceWith("primaryColor", "org.cru.godtools.tool.model.primaryColor"))
@get:ColorInt
inline val Styles?.primaryColor get() = primaryColor
@Deprecated(
    "Use kotlin mpp library instead",
    ReplaceWith("primaryTextColor", "org.cru.godtools.tool.model.primaryTextColor")
)
@get:ColorInt
inline val Styles?.primaryTextColor get() = primaryTextColor
@Deprecated("Use kotlin mpp library instead", ReplaceWith("textColor", "org.cru.godtools.tool.model.textColor"))
@get:ColorInt
inline val Styles?.textColor get() = textColor
val Styles?.textScale get() = this?.textScale ?: DEFAULT_TEXT_SCALE
@Deprecated("Use kotlin mpp library instead", ReplaceWith("textAlign", "org.cru.godtools.tool.model.textAlign"))
inline val Styles?.textAlign get() = textAlign

@Deprecated("Use kotlin mpp library instead", ReplaceWith("buttonStyle", "org.cru.godtools.tool.model.buttonStyle"))
inline val Styles?.buttonStyle get() = buttonStyle
